docs/proposals/20210722-yurtcluster-operator.md

```diff
@@ -72,43 +74,174 @@ This is an operator, with CRD and Controller.
+	// Operator will automatically override this value if it is less than the number of API Server.
+	// +optional
+	ServerCount int `json:"serverCount,omitempty"`
+	// PublicIP defines the public IP for tunnel server listen on.
+	// If this field is empty, the tunnel agent will use NodePort Service to connect to the tunnel server.
+	// +optional
+	PublicIP string `json:"publicIP,omitempty"`
+	// PublicPort defines the public port for tunnel server listen on.
+	// +optional
+	PublicPort int `json:"publicPort,omitempty"`
 }
 
-type FailedNode struct {
-        Name string `json:"name,omitempty"`
-        Info string `json:"info,omitempty"`
+// NodeSet defines a set of Kubernetes nodes.
+// It will merge the nodes that selected by Names, NamePattern, and Selector,
+// and then remove the nodes that match ExcludedNames and ExcludedNamePattern as the final set of nodes.
+type NodeSet struct {
+	// Names defines the node names to be selected
+	// +optional
+	Names []string `json:"names,omitempty"`
+	// NamePattern defines the regular expression to select nodes based on node name
+	// +optional
+	NamePattern string `json:"namePattern,omitempty"`
+	// Selector defines the label selector to select nodes
+	// +optional
+	Selector *corev1.NodeSelector `json:"selector,omitempty"`
+	// ExcludedNames defines the node names to be excluded
+	// +optional
+	ExcludedNames []string `json:"excludedNames,omitempty"`
+	// ExcludedNamePattern defines the regular expression to exclude nodes based on node name
+	// +optional
+	ExcludedNamePattern string `json:"excludedNamePattern,omitempty"`
 }
 
 // YurtClusterSpec defines the desired state of YurtCluster
 type YurtClusterSpec struct {
-        CloudNodes            NodeSpec        `json:"cloudNodes,omitempty"`
-        EdgeNodes             NodeSpec        `json:"edgeNodes,omitempty"`
-        YurtHub               ComponentConfig `json:"yurtHub,omitempty"`
-        YurtControllerManager ComponentConfig `json:"yurtControllerManager,omitempty"`
-        TunnelService         ComponentConfig `json:"yurtTunnel,omitempty"`
-        AppManager            ComponentConfig `json:"appManger,omitempty"`
+	// ImageRepository sets the container registry to pull images from.
+	// If empty, `docker.io/openyurt` will be used by default
+	// +optional
+	ImageRepository string `json:"imageRepository,omitempty"`
+	// YurtVersion is the target version of OpneYurt.
+	// +optional
+	YurtVersion string `json:"yurtVersion,omitempty"`
+	// CloudNodes defines the node set with cloud role.
+	// +optional
+	CloudNodes NodeSet `json:"cloudNodes,omitempty"`
+	// EdgeNodes defines the node set with edge role.
+	// +optional
+	EdgeNodes NodeSet `json:"edgeNodes,omitempty"`
+	// YurtHub defines the configuration for yurthub component.
+	// +optional
+	YurtHub YurtHub `json:"yurtHub,omitempty"`
+	// YurtTunnel defines the configuration for yurt tunnel component.
+	// +optional
+	YurtTunnel YurtTunnel `json:"yurtTunnel,omitempty"`
```

Comment:
I think we are able to use Helm chart to  deploy `YurtTunnel`, so `YurtTunnel` can be excluded in `YurtClusterSpec`

Reply:
If the proxy is implemented based on `iptables`, we need to keep `tunnel-server` and `kube-apiserver` deployed on the same node, operator is better suited to handle this. Based on `iptables`, there is no need to modify the configuration of `kube-apiserver`,  this make it easier for users to use `OpenYurt`. If the proxy is implemented based on DNS, we do not need to care about the deployment location of `tunnel-server`, but this will increase the complexity of user configuration, not only to modify the configuration of `kube-apiserver`, but also to modify the configuration of `CoreDNS`.  In order to make it easier for users, I recommend keeping `tunnel-server` can be deployed by the operator in `iptables` proxy mode, even if users can disable it and deploy it by helm chart.

Reply:
agree with your opinion.